File: partiql-ast/src/main/java/org/partiql/ast/v1/SetOp.java
package org.partiql.ast.v1;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Collections;

/**
 * TODO docs, equals, hashcode
 */
@Builder(builderClassName = "Builder")
@EqualsAndHashCode(callSuper = false)
public class SetOp extends AstNode {
    @NotNull
    public final SetOpType setOpType;

    @Nullable
    public final SetQuantifier setq;

    public SetOp(@NotNull SetOpType setOpType, @Nullable SetQuantifier setq) {
        this.setOpType = setOpType;
        this.setq = setq;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        return Collections.emptyList();
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitSetOp(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/SetOp.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class SetOp(
    @JvmField
    public var type: Type,
    @JvmField
    public var setq: SetQuantifier?,
) : AstNode() {
    public override fun children(): Collection<AstNode> = emptyList()

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = visitor.visitSetOp(
        this,
        ctx
    )

    /**
     * TODO docs, equals, hashcode
     */
    public enum class Type {
        UNION,
        INTERSECT,
        EXCEPT,
        OTHER,
    }
}
